documentation.md

```diff
@@ -0,0 +1,96 @@
+
+- Caller: The caller initiates the request to process a transaction.
+- Handle Transaction: This function handles the transaction, coordinates the processing steps, and accumulates the results.
+- Execute Request: This function processes the transaction for each typology and sends the results to CMS and CADP.
+- Evaluate Typology Expr: This function computes the typology result value based on the provided expression.
+- Redis Cache: Redis Cache stores interim rule results for faster processing and manages cache keys.
+- Database: The Database provides the typology expression needed for evaluation.
+- CMS: The CMS receives typology results that exceed the threshold.
+- CADP: The CADP receives all the typology results.
+
+## Sequence Diagram
+
+```mermaid
+sequenceDiagram
+    autonumber
+    participant Caller as Caller
+    participant HandleTransaction as Handle Transaction
+    participant ExecuteRequest as Execute Request
+    participant EvaluateTypologyExpr as Evaluate Typology Expr
+    participant RedisCache as Redis Cache
+    participant Database as Database
+    participant CMS as CMS
+    participant CADP as CADP
+
+    Note over Caller, HandleTransaction: The process starts by handling a transaction.
+
+    Caller->>HandleTransaction: call(transaction, networkMap, ruleResult)
+    HandleTransaction->>+ExecuteRequest: call(transaction, typology, ruleResult, networkMap)
+
+    Note over ExecuteRequest: Execute Request processes the transaction for each typology.
+
+    loop channel.typologies
+        ExecuteRequest->>RedisCache: addOneGetAll(cacheKey, ruleResult)
+        ExecuteRequest->>RedisCache: check cached ruleResults
+
+        Note over RedisCache: Redis Cache stores interim rule results for faster processing.
+
+        opt no cached ruleResults
+            ExecuteRequest->>HandleTransaction: return CADPRequest
+        end
+        ExecuteRequest->>Database: getTypologyExpression(typology)
+
+        Note over Database: Database provides the typology expression needed for evaluation.
+
+        ExecuteRequest->>EvaluateTypologyExpr: call(ruleValues, ruleResults, typologyExpression)
+
+        Note over EvaluateTypologyExpr: Evaluate Typology Expr computes the typology result value.
+
+        opt typologyResultValue > expression.threshold
+            ExecuteRequest->>CMS: send Typology result
+        end
+
+        Note over CMS: CMS receives typology results that exceed the threshold.
+
+        ExecuteRequest->>CADP: send Typology result
+
+        Note over CADP: CADP receives all the typology results.
+
+        ExecuteRequest->>RedisCache: deleteKey(cacheKey)
+
+        Note over RedisCache: Redis Cache deletes the interim cache key after processing.
+
+        ExecuteRequest->>HandleTransaction: return CADPRequest
+    end
+
+    HandleTransaction-->>ExecuteRequest: CADPRequest returned
+    HandleTransaction->>HandleTransaction: add CADPRequest to CombinedResult.cadpRequests
```

Comment:
Please confirm in the code? I think you are correct that it is happening inside the loop, but I suspect that it should not. The response from the typology processor to the calling rule processor only includes the result from one of the typologies (the first one it evaluates), and not all the typologies. If the typology processor responds to the rule processor from inside the loop, it would respond for every typology, which won't make sense because the rule processor will stop waiting for a response after the first response is received. It *should* wait until the loop is complete so that it can include with a response than includes ALL the typologies.